Move existing `tls-apps` to the same module as their topics

matthewhughes-uw opened the pull request:
So that teams have control over who produces to/consumes from their
topics. This follows on from 859c8b716d44e65733498b8cd67ba63f9304e9e8.
There should be no conflicting resources here, but we'll have to be
careful about applying to avoid deleting any existing ACLs, the plan is:

* `terraform state rm` in the module from which we removed the ACLs, so
  terraform no longer tracks them there
* `terraform apply` in the module we added the ACLs to, this should be a
  no-op as Terraform should detect the ACLs already exist and just add
  them to the state

Ticket: DENA-994

matthewhughes-uw commented:
I think this needs to be applied carefully: we don't want to delete the old ACLs before the new ones are created, but now they live in different modules

uw-infra commented:
Received terraform plan request
```
Cluster: dev-aws
Module: pubsub/kafka-shared-msk-account-identity
Path: dev-aws/kafka-shared-msk/account-identity
Commit ID: eabaf9be9d62bc73b8a4d0a471e11b969641058e
Requested At: 2024-10-17T09:37:05Z
```
Do not edit this comment. This message will be updated once the plan run is completed.
To manually trigger plan again please post `@terraform-applier plan dev-aws/kafka-shared-msk/account-identity` as comment.

uw-infra commented:
Received terraform plan request
```
Cluster: dev-aws
Module: pubsub/kafka-shared-msk-iam
Path: dev-aws/kafka-shared-msk/iam
Commit ID: eabaf9be9d62bc73b8a4d0a471e11b969641058e
Requested At: 2024-10-17T09:37:05Z
```
Do not edit this comment. This message will be updated once the plan run is completed.
To manually trigger plan again please post `@terraform-applier plan dev-aws/kafka-shared-msk/iam` as comment.

uw-infra commented:
Received terraform plan request
```
Cluster: dev-aws
Module: pubsub/kafka-shared-msk-cbc
Path: dev-aws/kafka-shared-msk/cbc
Commit ID: eabaf9be9d62bc73b8a4d0a471e11b969641058e
Requested At: 2024-10-17T09:37:06Z
```
Do not edit this comment. This message will be updated once the plan run is completed.
To manually trigger plan again please post `@terraform-applier plan dev-aws/kafka-shared-msk/cbc` as comment.

sbuliarca commented on dev-aws/kafka-shared-msk/cbc/cbc.tf:
```diff
@@ -1301,7 +1299,6 @@ module "cbc_legacy_account_eqdb_loader" {
   source           = "../../../modules/tls-app"
   produce_topics   = [kafka_topic.legacy_account_events_v2.name]
-  consume_topics   = ["account-identity.public.account.events"]
   consume_groups   = ["account-identity.cbc-account-events-relay-v2"]
```
I think it would be good to move also the related consumer group

matthewhughes-uw replied:
> I think it would be good to move also the related consumer group

is it not enough to have a single definition for the ACLs for those consumer groups (i.e. [here](https://github.com/utilitywarehouse/kafka-cluster-config/blob/d25341d70adae0880b323b7f83d666ea13f3357f/modules/tls-app/main.tf#L23)? Or would you want to duplicate them for visibility?

sbuliarca replied:
>is it not enough to have a single definition for the ACLs for those consumer groups

It would be enough. 
But consuming always happens within a consumer group. So while we moved the consume_topics to the module containing the topic, it would be reasonable to move also the consumer groups used for consuming from that topic.

sbuliarca commented on dev-aws/kafka-shared-msk/cbc/cbc.tf:
```diff
@@ -365,8 +365,6 @@ module "cbc_fraud_detection_api" {
 module "cbc_fraud_detection_consumer" {
   source = "../../../modules/tls-app"
   consume_topics = [
-    "auth-customer.iam-credentials-v1-public",
```
Same for the related consumer groups.

sbuliarca commented:
> I think this needs to be applied carefully: we don't want to delete the old ACLs before the new ones are created, but now they live in different modules

Indeed. How we can approach this is: before merging the PR, remove from the tf state the ACLs that would be deleted. This way, on apply TF will not remove them. 
But we need to do it manually through the https://developer.hashicorp.com/terraform/cli/commands/state/rm command. 
AFAIK we are safe on the creation side, as TF will not fail if the ACL already exists, but it will only add it to the tracked state.

uw-infra commented:
Received terraform plan request
```
Cluster: dev-aws
Module: pubsub/kafka-shared-msk-account-identity
Path: dev-aws/kafka-shared-msk/account-identity
Commit ID: 661481900dda75ba7fa07e11294c5307118d7a84
Requested At: 2024-10-17T13:16:09Z
```
Do not edit this comment. This message will be updated once the plan run is completed.
To manually trigger plan again please post `@terraform-applier plan dev-aws/kafka-shared-msk/account-identity` as comment.

uw-infra commented:
Received terraform plan request
```
Cluster: dev-aws
Module: pubsub/kafka-shared-msk-cbc
Path: dev-aws/kafka-shared-msk/cbc
Commit ID: 661481900dda75ba7fa07e11294c5307118d7a84
Requested At: 2024-10-17T13:16:10Z
```
Do not edit this comment. This message will be updated once the plan run is completed.
To manually trigger plan again please post `@terraform-applier plan dev-aws/kafka-shared-msk/cbc` as comment.

matthewhughes-uw commented:
> Indeed. How we can approach this is: before merging the PR, remove from the tf state the ACLs that would be deleted. This way, on apply TF will not remove them.
But we need to do it manually through the https://developer.hashicorp.com/terraform/cli/commands/state/rm command.
AFAIK we are safe on the creation side, as TF will not fail if the ACL already exists, but it will only add it to the tracked state.

Are you sure it won't just create _new_ ACLs with exactly the same settings, but a different ID? That's what the current plan output looks to me

uw-infra commented:
Received terraform plan request
```
Cluster: dev-aws
Module: pubsub/kafka-shared-msk-iam
Path: dev-aws/kafka-shared-msk/iam
Commit ID: 4444c4f9cb65da89998e23e5a4ee093a63640b06
Requested At: 2024-10-18T07:57:54Z
```
Do not edit this comment. This message will be updated once the plan run is completed.
To manually trigger plan again please post `@terraform-applier plan dev-aws/kafka-shared-msk/iam` as comment.

uw-infra commented:
Received terraform plan request
```
Cluster: dev-aws
Module: pubsub/kafka-shared-msk-account-identity
Path: dev-aws/kafka-shared-msk/account-identity
Commit ID: 4444c4f9cb65da89998e23e5a4ee093a63640b06
Requested At: 2024-10-18T07:57:54Z
```
Do not edit this comment. This message will be updated once the plan run is completed.
To manually trigger plan again please post `@terraform-applier plan dev-aws/kafka-shared-msk/account-identity` as comment.

Nesze commented on dev-aws/kafka-shared-msk/iam/iam.tf:
```diff
@@ -292,3 +292,9 @@ module "castle_processor" {
   consume_groups   = ["iam.castle-processor"]
 }
+
+module "cbc_fraud_detection_consumer" {
+  source           = "../../../modules/tls-app"
+  consume_topics   = [kafka_topic.iam_credentials_v1_public.name]
+  cert_common_name = "cbc/cbc-fraud-detection-consumer"
+}
```
does this also need a consume group?

sbuliarca replied:
Indeed, I also checked this, and they use the same consumer group for multiple topics: https://kafka-ui.dev.uw.systems/ui/clusters/shared-msk/consumer-groups/cbc.cbc-fraud-detection-consumer-v1
The ACL for it needs to be defined in a single place

Nesze replied:
thanks, make sense 👍

matthewhughes-uw commented:
Dropped those resources from `cbc`'s state:

```console
$ terraform plan -out tf-plan
$ terraform show -json tf-plan | jq --raw-output '.resource_changes[] | select(.change.actions[] | index("delete")) | .address' | xargs -I {} terraform state rm '{}'
Removed module.cbc_account_events_relay.kafka_acl.group_acl["account-identity.cbc-account-events-relay-v2"]
Successfully removed 1 resource instance(s).
Removed module.cbc_account_events_relay.kafka_acl.topic_acl["account-identity.public.account.events"]
Successfully removed 1 resource instance(s).
Removed module.cbc_fraud_detection_consumer.kafka_acl.group_acl["account-identity.cbc-fraud-detection-consumer-v1"]
Successfully removed 1 resource instance(s).
Removed module.cbc_fraud_detection_consumer.kafka_acl.topic_acl["account-identity.public.account.events"]
Successfully removed 1 resource instance(s).
Removed module.cbc_fraud_detection_consumer.kafka_acl.topic_acl["auth-customer.iam-credentials-v1-public"]
Successfully removed 1 resource instance(s).
```

I'll merge this PR and let `kube-applier` handle the creation in `iam` and `account-identity` then turn off `planOnly` in `cbc`

uw-infra commented:
Terraform run output for
```
Cluster: dev-aws
Module: pubsub/kafka-shared-msk-iam
Path: dev-aws/kafka-shared-msk/iam
Commit ID: cd7c0f097ab7fc6cedd48a93a70a9ccae27ee7e2
```
<details><summary><b>✅ Run Status: Ok, Run Summary: Apply complete! Resources: 1 added, 0 changed, 0 destroyed</b></summary>

```terraform

Terraform used the selected providers to generate the following execution
plan. Resource actions are indicated with the following symbols:
  + create

Terraform will perform the following actions:

  # module.cbc_fraud_detection_consumer.kafka_acl.topic_acl["auth-customer.iam-credentials-v1-public"] will be created
  + resource "kafka_acl" "topic_acl" {
      + acl_host                     = "*"
      + acl_operation                = "Read"
      + acl_permission_type          = "Allow"
      + acl_principal                = "User:CN=cbc/cbc-fraud-detection-consumer"
      + id                           = (known after apply)
      + resource_name                = "auth-customer.iam-credentials-v1-public"
      + resource_pattern_type_filter = "Literal"
      + resource_type                = "Topic"
    }

Plan: 1 to add, 0 to change, 0 to destroy.
module.cbc_fraud_detection_consumer.kafka_acl.topic_acl["auth-customer.iam-credentials-v1-public"]: Creating...
module.cbc_fraud_detection_consumer.kafka_acl.topic_acl["auth-customer.iam-credentials-v1-public"]: Creation complete after 1s [id=User:CN=cbc/cbc-fraud-detection-consumer|*|Read|Allow|Topic|auth-customer.iam-credentials-v1-public|Literal]

Warning: Argument is deprecated

  with provider["registry.terraform.io/mongey/kafka"],
  on __env.tf line 12, in provider "kafka":
  12: provider "kafka" {

This parameter is now deprecated and will be removed in a later release,
please use `client_cert` instead.

(and one more similar warning elsewhere)

Apply complete! Resources: 1 added, 0 changed, 0 destroyed.

```
</details>

> To manually trigger plan again please post `@terraform-applier plan dev-aws/kafka-shared-msk/iam` as comment.

uw-infra commented:
Terraform run output for
```
Cluster: dev-aws
Module: pubsub/kafka-shared-msk-cbc
Path: dev-aws/kafka-shared-msk/cbc
Commit ID: cd7c0f097ab7fc6cedd48a93a70a9ccae27ee7e2
```
<details><summary><b>✅ Run Status: Ok, Run Summary: No changes. Your infrastructure matches the configuration.</b></summary>

```terraform

No changes. Your infrastructure matches the configuration.

Terraform has compared your real infrastructure against your configuration
and found no differences, so no changes are needed.

```
</details>

> To manually trigger plan again please post `@terraform-applier plan dev-aws/kafka-shared-msk/cbc` as comment.

uw-infra commented:
Terraform run output for
```
Cluster: dev-aws
Module: pubsub/kafka-shared-msk-account-identity
Path: dev-aws/kafka-shared-msk/account-identity
Commit ID: cd7c0f097ab7fc6cedd48a93a70a9ccae27ee7e2
```
<details><summary><b>✅ Run Status: Ok, Run Summary: Apply complete! Resources: 4 added, 0 changed, 0 destroyed</b></summary>

```terraform

Terraform used the selected providers to generate the following execution
plan. Resource actions are indicated with the following symbols:
  + create

Terraform will perform the following actions:

  # module.cbc_account_events_relay.kafka_acl.group_acl["account-identity.cbc-account-events-relay-v2"] will be created
  + resource "kafka_acl" "group_acl" {
      + acl_host                     = "*"
      + acl_operation                = "Read"
      + acl_permission_type          = "Allow"
      + acl_principal                = "User:CN=cbc/cbc-account-events-relay-v2"
      + id                           = (known after apply)
      + resource_name                = "account-identity.cbc-account-events-relay-v2"
      + resource_pattern_type_filter = "Literal"
      + resource_type                = "Group"
    }

  # module.cbc_account_events_relay.kafka_acl.topic_acl["account-identity.public.account.events"] will be created
  + resource "kafka_acl" "topic_acl" {
      + acl_host                     = "*"
      + acl_operation                = "Read"
      + acl_permission_type          = "Allow"
      + acl_principal                = "User:CN=cbc/cbc-account-events-relay-v2"
      + id                           = (known after apply)
      + resource_name                = "account-identity.public.account.events"
      + resource_pattern_type_filter = "Literal"
      + resource_type                = "Topic"
    }

  # module.cbc_fraud_detection_consumer.kafka_acl.group_acl["account-identity.cbc-fraud-detection-consumer-v1"] will be created
  + resource "kafka_acl" "group_acl" {
      + acl_host                     = "*"
      + acl_operation                = "Read"
      + acl_permission_type          = "Allow"
      + acl_principal                = "User:CN=cbc/cbc-fraud-detection-consumer"
      + id                           = (known after apply)
      + resource_name                = "account-identity.cbc-fraud-detection-consumer-v1"
      + resource_pattern_type_filter = "Literal"
      + resource_type                = "Group"
    }

  # module.cbc_fraud_detection_consumer.kafka_acl.topic_acl["account-identity.public.account.events"] will be created
  + resource "kafka_acl" "topic_acl" {
      + acl_host                     = "*"
      + acl_operation                = "Read"
      + acl_permission_type          = "Allow"
      + acl_principal                = "User:CN=cbc/cbc-fraud-detection-consumer"
      + id                           = (known after apply)
      + resource_name                = "account-identity.public.account.events"
      + resource_pattern_type_filter = "Literal"
      + resource_type                = "Topic"
    }

Plan: 4 to add, 0 to change, 0 to destroy.
module.cbc_fraud_detection_consumer.kafka_acl.group_acl["account-identity.cbc-fraud-detection-consumer-v1"]: Creating...
module.cbc_account_events_relay.kafka_acl.group_acl["account-identity.cbc-account-events-relay-v2"]: Creating...
module.cbc_fraud_detection_consumer.kafka_acl.topic_acl["account-identity.public.account.events"]: Creating...
module.cbc_account_events_relay.kafka_acl.topic_acl["account-identity.public.account.events"]: Creating...
module.cbc_account_events_relay.kafka_acl.topic_acl["account-identity.public.account.events"]: Creation complete after 0s [id=User:CN=cbc/cbc-account-events-relay-v2|*|Read|Allow|Topic|account-identity.public.account.events|Literal]
module.cbc_fraud_detection_consumer.kafka_acl.topic_acl["account-identity.public.account.events"]: Creation complete after 0s [id=User:CN=cbc/cbc-fraud-detection-consumer|*|Read|Allow|Topic|account-identity.public.account.events|Literal]
module.cbc_fraud_detection_consumer.kafka_acl.group_acl["account-identity.cbc-fraud-detection-consumer-v1"]: Creation complete after 0s [id=User:CN=cbc/cbc-fraud-detection-consumer|*|Read|Allow|Group|account-identity.cbc-fraud-detection-consumer-v1|Literal]
module.cbc_account_events_relay.kafka_acl.group_acl["account-identity.cbc-account-events-relay-v2"]: Creation complete after 0s [id=User:CN=cbc/cbc-account-events-relay-v2|*|Read|Allow|Group|account-identity.cbc-account-events-relay-v2|Literal]

Warning: Argument is deprecated

  with provider["registry.terraform.io/mongey/kafka"],
  on __env.tf line 12, in provider "kafka":
  12: provider "kafka" {

This parameter is now deprecated and will be removed in a later release,
please use `client_cert` instead.

(and one more similar warning elsewhere)

Apply complete! Resources: 4 added, 0 changed, 0 destroyed.

```
</details>

> To manually trigger plan again please post `@terraform-applier plan dev-aws/kafka-shared-msk/account-identity` as comment.
